### DISCUSSION TITLE
Customisability of the abp Commercial Team Edition

### DISCUSSION THREAD

**Discussion by KevinFarrow** [abpframework/abp#6246] [General]
Great product abp.io but we have some questions...

We are looking at purchasing a copy of abp commercial team edition, but we have a few questions that we would like answering before we buy it. I think there will be many people in this thread that have similar questions and I believe it would help everyone if you could answer these few question to give us a clearer picture of how customisable the abp commercial team edition is?

1. What source code do we get with the Team edition? HTML, css, Javascript, C# etc. We understand you don't want to give away the C# code in this edition but we would need to at least change some of the css and html.

2. I know someone has posted an article and html source for the login page to allow us to change it, but will this code be always available? If you make any breaking changes, will the latest html be available for us to change the login page?
	
3. If you don't give any of the HTML or css how do we change the logo at the top left of every page for each Tenant? Looking at the source of the demo, the image is called bg-transparent.32194ca4d410195f69d6.png - do we have to replace that image with a same named .png image? and every Tenant will have that logo?
	
4. If we don't have the html source, how do we change the default Dashboard? or the Tenant Dashboard? How do we add our own widgets and change the layout?
	
5. How do we change the footer message? On the demo it says "2019 - 2020 © AbpCommercialDemo" - can we change this? Or possibly remove the footer component?
	
6. At some point I would think you will be adding notifications. If you do, how would we be able to add our own notifications without the source code? Could we inherit and override existing classes? Will you provide details of the methods we would be able to override?

7. Last but not least, do the Community posts refer to the open source version or the Team version or the Business edition? It would help if you added a label to the posts as it is difficult to know what post works with what version. For example the post about changing the theme to AdminLTE is probably only applicable to the open source version.

As you see we have many questions and cannot afford $5,000 for full source code as that is very expensive to us, so if we could change just a few things and override some of the classes that would be great. Thanks for any help.

**Comment by hikalkan** [comment]
Hi,

To be a reference to everyone, I will answer the questions in details.

With the [ASP.NET Zero](https://aspnetzero.com/), we were giving the **full source code**, so you can change however you like. Then we release a new version with new features, improvements and bugfixes, but **you can't easily upgrade** your solution to get the latest code; Because you changed the source code and you need to perform a merge operation which may not be easy always. We got a lot of feedback from customers, but there was no easy way to do it as long as we deliver the full source code. That's one of the major reasons we had to re-write the entire platform.

As a solution, now the startup template is **almost empty** and all the pre-built functionalies come with **reusable, upgradeable, removable NuGet & NPM packages**. You can upgrade your solution with a single command: [abp update](https://docs.abp.io/en/abp/latest/Upgrading).

Every decision comes with a cost. The cost here is **customizability**. We were aware of that from the first day and provided a lot of different ways to customize the pre-built modules. Now, I believe **it has enough customizability/extensibility options**. Most of these options are **already documented** and we continue to write documents and tutorials. You can start from the following docs:

* https://docs.abp.io/en/abp/latest/Customizing-Application-Modules-Guide (open source side)
* https://docs.abp.io/en/commercial/latest/guides/customizing-modules (commercial side - most of these options will be available to open source side too in a short time (actually they are available but we haven't moved the docs yet)).

ABP Commercial Team license owners can use all the modules, themes and tools available to Business/Enterprise license owners. The main difference is that Business/Enterprise license owners can download the source code of any module/theme, and if they want, they can move the source code into their own solution for maximum level of customization (while they loose auto upgrade  possibility for that module - it is a tradeoff). Even if they don't copy the code to their solution, having the source code makes easier to understand the code structure and customize/override it on need (and partially copy & customize).

For Team License owners, they can't see the fully source code which can be a problem especially for the UI side (because they can inherit a C# class and override it easily). So, if you need to fully see the source code, you have to buy an Enterprise or Business license. However, we are trying to document internals of all the modules, so most of the times you don't need to see its source code.

We understand that $4,999 can be expensive for some companies. But remember that we are developing the abp.io (including the open source side that we don't make any money) more than 4 years with a lot of professional developers and it has a big cost to us 😃 I think the time and money saving is a lot more than $4,999 for a company that builds a serious product. BTW, ABP Commercial licenses allow you to develop unlimited of different products with a single license.

After all the story, I can answer to your questions;

1. What you get is almost identical to the solution when you create with the free startup templates. The difference is NuGet/NPM packages (for modules & the theme) and some configuration related to these modules.

2. For Business/Enterprise licenses, you can always download the latest source code of any module.

3. This [document](https://docs.abp.io/en/abp/latest/UI/AspNetCore/Customization-User-Interface#overriding-a-view-component) shows how to replace a UI component. It is similar to the ABP Commercial. So, you can replace the Branding component. Also, [there is an easier way ](https://docs.abp.io/en/abp/latest/UI/AspNetCore/Branding) for the Branding part. We continue to improve the documents.

4. Dashboards are in the startup template and not a part of the NuGet packages. So, you can freely change them. We use widget system to define widgets on the dashboard. A module may define a reusable widget that you can directly use in your dashboard and it becomes upgradable even the dashboard code is in your solution.

5. Footer is a view component and already overriden in the startup template. So, you can freely change it in the startup template.

6. I didn't understand much. You can build your own notification system. If you want to put a notification icon to layout, there is an exact [example here](https://docs.abp.io/en/abp/latest/UI/AspNetCore/Customization-User-Interface#toolbar-contributors).

7. Most of the posts refer to the open source version. Everything for the open source side is already valid for the ABP Commercial. Some posts are for the ABP Commercial and we indicate that at the beginning of the post. We will use tags in the future but it is not a problem now, believe me. AdminLTE is applicable to the ABP Commercial too. The [Lepton Theme](https://commercial.abp.io/themes) is also replaceable and none of the ABP Commercial modules depend on the Lepton theme. [Theming system](https://docs.abp.io/en/abp/latest/UI/AspNetCore/Theming) is really powerful with the ABP Framework.

Finally,

I suggest the Team License if you have limited budget, you won't make major changes in the existing modules and upgradability is more important for you.

I suggest Business/Enteprise License;

* If you want to get the full source code of the modules & themes. This is not only for customization. Enterprise companies find it safer for themselves to have the full code of the product.
* If email/private support is important (these licenses have also more support question limit - enteprise is limitless).
